pkg/clients/postgresql/postgresql.go

```diff
@@ -3,38 +3,76 @@ package postgresql
 import (
 	"context"
 	"database/sql"
+	"fmt"
 
 	"github.com/crossplane-contrib/provider-sql/pkg/clients/xsql"
 
 	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
 	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
 )
 
+const (
+	panicNilPtr = "function call on nil pointer"
+)
+
 type postgresDB struct {
 	dsn      string
+	username string
+	password string
+	database string
 	endpoint string
 	port     string
 }
 
 // New returns a new PostgreSQL database client.
+// The default database name is an empty string.
+// The underlying pq library will default to either
+// using the value of PGDATABASE, or if unset, the
+// hardcoded db name 'postgres'.
+// The db name can be overridden later using SetDatabase,
+// which will trigger a re-format of the DSN.
 func New(creds map[string][]byte) xsql.DB {
 	// TODO(negz): Support alternative connection secret formats?
-	endpoint := string(creds[xpv1.ResourceCredentialsSecretEndpointKey])
-	port := string(creds[xpv1.ResourceCredentialsSecretPortKey])
-	return postgresDB{
-		dsn: "postgres://" +
-			string(creds[xpv1.ResourceCredentialsSecretUserKey]) + ":" +
-			string(creds[xpv1.ResourceCredentialsSecretPasswordKey]) + "@" +
-			endpoint + ":" +
-			port,
-		endpoint: endpoint,
-		port:     port,
+	db := postgresDB{
+		username: string(creds[xpv1.ResourceCredentialsSecretUserKey]),
+		password: string(creds[xpv1.ResourceCredentialsSecretPasswordKey]),
+		endpoint: string(creds[xpv1.ResourceCredentialsSecretEndpointKey]),
+		port:     string(creds[xpv1.ResourceCredentialsSecretPortKey]),
 	}
+	db.formatDSN()
+	return &db
+}
+
+// formatDSN sets the DSN that this client will use
+// to connect to PostgreSQL.
+func (c *postgresDB) formatDSN() {
+	if c == nil {
+		panic(panicNilPtr)
+	}
+	c.dsn = fmt.Sprintf("postgres://%s:%s@%s:%s/%s",
+		c.username,
+		c.password,
+		c.endpoint,
+		c.port,
+		c.database)
+}
+
+// SetDatabase overrides the existing database name,
+// and regenerates the stored DSN.
+func (c *postgresDB) SetDatabase(db string) {
+	if c == nil {
+		panic(panicNilPtr)
+	}
+	c.database = db
+	c.formatDSN()
 }
 
 // ExecTx executes an array of queries, committing if all are successful and
 // rolling back immediately on failure.
-func (c postgresDB) ExecTx(ctx context.Context, ql []xsql.Query) error {
+func (c *postgresDB) ExecTx(ctx context.Context, ql []xsql.Query) error {
+	if c == nil {
```

Comment:
As mentioned above; I don't think it's necessary to add these `c == nil` checks.